### PR TITLE
Issue 1386: Smart groups - list view

### PR DESCRIPTION
**Pull request by dleszcz**
**Description**

Main:

- Place for groups in Redux store
- List item (UI) + Fetching groups from BE
- Filters by name/description
- Select / Multiselect group items

Menu:

- Colouring all groups 
- Removing all groups
- Download JSON file

Item Actions:

- Colouring group 
- Removing group
- Isolate group

Footer:

- add button
- counting groups in footer

### CODE REVIEW

File: frontend/modules/groups/groups.sagas.ts [hunk 1]
```diff
@@ -15,19 +15,253 @@
...
+
+		if (treeMap) {
+			const nodes = yield TreeService.getNodesFromSharedIds(group.objects);
+
+			if (nodes) {
+				const filteredNodes = nodes.filter((n) => n !== undefined);
+				const meshes = TreeService.getMeshMapFromNodes(filteredNodes);
+
+				for (const key in meshes) {
+					if (key) {
```
Comment (simplypixi) [inline frontend/modules/groups/groups.sagas.ts]: Is it necessary? I think it'll be always true. :)

File: frontend/modules/groups/groups.sagas.ts [hunk 2]
```diff
@@ -15,19 +15,253 @@
...
+						const meshIds = meshes[key].meshes;
+						const pair = key.split('@');
+						const modelAccount = pair[0];
```
Comment (simplypixi) [inline frontend/modules/groups/groups.sagas.ts]:
```suggestion
						const [modelAccount, modelId] = key.split('@');
```

File: frontend/modules/groups/groups.sagas.ts [hunk 3]
```diff
@@ -15,19 +15,253 @@
...
+	} catch (error) {
+		yield put(DialogActions.showErrorDialog('color', 'override', error));
+	}
+}
+
+export function* removeColorOverride({ groupId, overridedGroup }) {
+	try {
+		if (overridedGroup) {
+			for (const key in overridedGroup.models) {
+				if (overridedGroup.models.hasOwnProperty(key)) {
```
Comment (simplypixi) [inline frontend/modules/groups/groups.sagas.ts]: like above

File: frontend/modules/groups/groups.sagas.ts [hunk 4]
```diff
@@ -15,19 +15,253 @@
...
+					const meshIds = overridedGroup.models[key].meshes;
+					const pair = key.split('@');
+					const account = pair[0];
```
Comment (simplypixi) [inline frontend/modules/groups/groups.sagas.ts]: like above

File: frontend/modules/groups/groups.sagas.ts [hunk 5]
```diff
@@ -15,19 +15,253 @@
...
+
+export function* deleteGroups({ teamspace, modelId, groups }) {
+	try {
+		yield API.deleteGroups(teamspace, modelId, groups);
+
+		const groupsToDelete = groups.split(',');
+		const colorOverrides = yield select(selectColorOverrides);
+		const groupsMap = yield select(selectGroupsMap);
+
+		yield all(groupsToDelete.map((groupId) => {
```
Comment (simplypixi) [inline frontend/modules/groups/groups.sagas.ts]: Does `all` support nested arrays?

File: frontend/routes/components/filterPanel/filterPanel.component.tsx [hunk 6]
```diff
@@ -63,9 +64,10 @@ interface ISelectedFilter {
 }
 
 interface IProps {
-	filters: IFilter[];
+	filters?: IFilter[];
 	onChange: (selectedFilters) => void;
 	selectedFilters: any[];
+	hideFiltersMenu?: boolean;
```
Comment (simplypixi) [inline frontend/routes/components/filterPanel/filterPanel.component.tsx]: You no need to add `Filters` into the name. It is filtersPanel, so it is logical that this prop is related to it. :P

File: frontend/routes/components/filterPanel/filterPanel.component.tsx [hunk 7]
```diff
@@ -63,9 +64,10 @@ interface ISelectedFilter {
 }
 
 interface IProps {
-	filters: IFilter[];
+	filters?: IFilter[];
```
Comment (simplypixi) [inline frontend/routes/components/filterPanel/filterPanel.component.tsx]: 👍

File: frontend/routes/components/fontAwesomeIcon/fontAwesomeIcon.component.tsx [hunk 8]
```diff
@@ -25,11 +25,13 @@ interface IExtendedIconProps extends IconProps {
 }
 interface IProps {
 	icon: any;
+	size?: any;
```
Comment (simplypixi) [inline frontend/routes/components/fontAwesomeIcon/fontAwesomeIcon.component.tsx]: What about `IconProps`?

File: frontend/routes/viewer/components/groups/components/groupsListItem/groupsListItem.component.tsx [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import { getGroupRGBAColor } from './../../../../../../helpers/colors';
+import { GROUPS_TYPES } from './../../../../../../helpers/groups';
+import { DEFAULT_OVERRIDE_COLOR } from './../../../../../../constants/groups';
+import { DateTime } from '../../../../../components/dateTime/dateTime.component';
+import { renderWhenTrue } from '../../../../../../helpers/rendering';
+import { TooltipButton } from '../../../../../teamspaces/components/tooltipButton/tooltipButton.component';
+
+interface IProps {
+	_id: string;
+	autohr: string;
```
Comment (simplypixi) [inline frontend/routes/viewer/components/groups/components/groupsListItem/groupsListItem.component.tsx]: :D

File: frontend/routes/viewer/components/groups/groups.component.tsx [hunk 10]
```diff
@@ -44,18 +158,114 @@ export class Groups extends React.PureComponent<IProps, any> {
 		return <GroupWork />;
 	}
 
+	get menuActionsMap() {
+		const { groups, deleteGroups, toggleColorOverrideAll, teamspace, model, downloadGroups } = this.props;
+		return {
+			[GROUPS_ACTIONS_ITEMS.OVERRIDE_ALL]: () => toggleColorOverrideAll(),
+			[GROUPS_ACTIONS_ITEMS.DELETE_ALL]: () => {
+				this.props.showConfirmDialog({
```
Comment (simplypixi) [inline frontend/routes/viewer/components/groups/groups.component.tsx]: Could you move this to a new handler?

File: frontend/services/viewer/viewer.ts [hunk 11]
```diff
@@ -202,6 +202,22 @@ export class ViewerService {
 			this.viewer.stopAreaSelect();
 		}
 	}
+
+	public getDefaultHighlightColor() {
+		if (this.viewer) {
```
Comment (simplypixi) [inline frontend/services/viewer/viewer.ts]: Is it necessary?

File: frontend/modules/groups/groups.sagas.ts (hunk 5 above)
Comment (dleszcz) [reply frontend/modules/groups/groups.sagas.ts]: Yes, it works fine

**Pull request merged**